diff --git a/Makefile b/Makefile
index 0219d66..bbeb63a 100644
--- a/Makefile
+++ b/Makefile
@@ -10,12 +10,13 @@ SRC := $(wildcard src/*.cpp)
 
 SPIKE?=../riscv-isa-sim
 SPIKE_BUILD=$(realpath ${SPIKE}/build)
+ELFIO_BUILD=$(realpath ${SPIKE}/include)
 SPIKE_OBJS:= libspike_main.a  libriscv.a  libdisasm.a  libsoftfloat.a  libfesvr.a  libfdt.a
 SPIKE_OBJS:=$(addprefix ${SPIKE_BUILD}/,${SPIKE_OBJS})
 LDFLAGS+=${SPIKE_OBJS}
 LDFLAGS += -L/usr/lib/x86_64-linux-gnu
 LIBRARIES += -lpthread -ldl -lboost_regex -lboost_system -lpthread  -lboost_system -lboost_regex 
-
+INCLUDE += -I$(realpath ${ELFIO_BUILD})
 INCLUDE += -I$(realpath ${SPIKE}/riscv)
 INCLUDE += -I$(realpath ${SPIKE}/fesvr)
 INCLUDE += -I$(realpath ${SPIKE}/softfloat)
